ArcGISAppBuilderWidget/README.md

```diff
@@ -0,0 +1,162 @@
+> `widgetPool` widgets are widgets that appear in the top level controller of
+> the application.  These typically include the Legend and Layer List widgets.
+
+1. Verify [Web AppBuilder For ArcGIS version
+   `2.16`](https://developers.arcgis.com/web-appbuilder/guide/getstarted.htm) is
+   being used.
+2. Copy contents of [./dist](./dist) into your `/widgets` dir of your app's directory.
+3. Edit the `config.json` in the root of the app’s directory, and in the
+   `widgetPool.widgets` array, add this definition:
+
+    ```json
+    {
+      "name": "OptiSites",
+      "version": "2.16",
+      "IsController": false,
+      "uri": "widgets/OptiSites/Widget",
+      "config": "configs/OptiSites/config_widgets_OptiSites_Widget_20.json",
+      "id": "widgets_OptiSites_Widget_20",
+      "index": 7,
+      "openAtStart": true
+    }
+    ```
+4. Add `configs/OptiSites/config_widgets_OptiSites_Widget_20.json`
+
+   ```json
+   {
+     "optiAPIKey": "Your public key here",
+     "zoomToFitOptiSites": true
+   }
+   ```
+
+## Notes about dependencies
+
+### Dependencies:
+
+In addition to [AMD modules](https://github.com/amdjs/amdjs-api) provided by
+ArcGIS's AppBuilder, this example widget has 2 dependencies (non development dependencies):
+
+* [cross-fetch](https://github.com/lquixada/cross-fetch): A WHATWG Fetch API
+  for older environments that do not support
+  [`fetch()`](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API)
+* [dygraphs](https://github.com/danvk/dygraphs): A libray for rendering time
+  series charts of data from Opti's public API.
+
+These dependencies are included in [./OptiSites/js](./OptiSites/js) folder and
+this is where the widget imports these dependencies from. It is the general practice
+for ArcGIS AppBuilder widgets to import 3rd party vendor scripts this way.
+
+These dependencies are also specified in the [package.json](./package.json) and installed by
+`npm install`. The content installed by `npm` into `node_modules` is only used for development.
+Some use cases for copies in `npm_modules` include typescript definitions, `npm outdated` & `npm audit`,
+or to inspect their licenses.  If upgraded using `npm`, be sure to update the applicable scripts in
+[./OptiSites/js](./OptiSites/js) because that's where AppBuilder's widget imports them.
+
+### Development Dependencies:
+
+Other `devDependencies` are specified in the [package.json](./package.json) and
+installed with `npm install`.
+
+Notes about `devDependencies`:
+* includes typescript definitions
+* ArcGIS AppBuilder uses version 3 of the `arcgis-js-api`, and not the more
+  recent version 4.
+  > Do not upgrade the type definitions `@types/arcgis-js-api` to v4.x because
+  > there are breaking changes in v4.
+* [momentjs](https://momentjs.com) is included in ArcGIS's AppBuilder.  It is
+  available for `import` into the widget source.  The only reason for including
+  it as a `devDependency` is for its type definitions.
+
+## Additional Developer Notes:
```

Comment:
This is valuable advice. Are there resources for "Typescript development for ArcGIS AppBuilder 2.16" or is this worth publishing in a blog context to a more general web development audience?

Reply:
Documentation on ArcGIS's AppBuilder and Typescript is lacking.  Their newer v4 JS api is better because they have adopted Typescript.  But AppBuilder is based on the v3 JS api.

I found these docs to be the best references.  But I spent a lot of time studying these to figure out the APIs in a modern JS / Typescript context.

* Reading the type definitions of `@types/arcgis-js-api@3` (version 3)
* https://developers.arcgis.com/javascript/3/jsapi/
* https://developers.arcgis.com/web-appbuilder/guide/getstarted.htm
* https://developers.arcgis.com/web-appbuilder/api-reference/css-framework.htm
* https://developers.arcgis.com/web-appbuilder/sample-code/create-custom-in-panel-widget.htm